Allow user to set log4j logging level.

cmnbroad opened the pull request:
This PR dynamically sets the logging level for command line tools at runtime using the current version of log4j (we were headed down a path of downgrading to a previous version of log4j in order to implement this). However, it uses an API that is normally used in code for extending log4j rather than acting as a client to it, and requires an explicit cast of the value returned from LogManager.getContext. The Apache project site illustrates the use of this api in the first line of code in an example here: https://logging.apache.org/log4j/2.x/manual/customconfig.html#AddingToCurrent.

We need to decide if we want to take this and stay on the current version or continue with the downgrade…

akiezun commented:
@droazen or @lbergelson can you review it? If not, I'll have a look on Wednesday

lbergelson commented on src/test/java/org/broadinstitute/hellbender/MainTest.java:
```diff
@@ -16,4 +20,14 @@ public void testPrintReadsVersion() {
+    public void testCustomLogLevel() {
+        // This first assert only serves to validate the test conditions and ensure
+        // that the rest of the test is valid.
+        Assert.assertTrue(logger.getLevel() != Level.DEBUG);
```
Thanks for this comment.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/MainTest.java:
```diff
@@ -16,4 +20,14 @@ public void testPrintReadsVersion() {
+        Assert.assertTrue(logger.getLevel() != Level.DEBUG);
+        final File testFile = new File(getTestDataDir(), "print_reads.sam");
+        Main.main(new String[]{"CountReads", "-I", testFile.getPath(), "--VERBOSITY", "DEBUG"});
+        Assert.assertTrue(logger.getLevel() == Level.DEBUG);
```
So this is setting the global logger, but I'm not sure what that means for the log level of the rest of the tests.  Would it be possible to capture the initial logging level and the reset it after the test?

cmnbroad replied:
Yeah. Good catch...

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -162,6 +168,38 @@ public Object instanceMain(final String[] argv) {
     /**
+     * Propagate the verbosity level to both the picard and log4j loggers.
+     */
+    protected void setLoggingLevel(final Log.LogLevel VERBOSITY) {
```
Could this method be made static and moved into Utils?  I think it makes more sense as a static method since it sets global static state rather than anything specific to this class.

lbergelson commented:
@cmnbroad Sorry for the very long delay before reviewing this.  I think this is a good solution.  I think the test is a bit weird at the moment, but it's a weird thing to try and test.  

1: Could you make `setLoggingLevel` a static method in `Utils`, and give it a clear comment saying that it changes the global logging level (it does change the global logging level yes?  Do I misunderstand what's happening? )  

2:  Could you then move the test to `UtilsUnitTest` run through all 4 levels of logging in the test, and then reset it to the initial state (if that's possible..) 

Thanks for figuring this out and sorry for delay.

cmnbroad commented:
Just to add some clarity around the the granularity of these log levels: the setLoggingLevel API is setting the level of all of the loggers that inherit from the caller's context configuration, which as I understand it, is essentially global for Hellbender, since it currently has only one configuration.

cmnbroad commented:
I moved and rewrote the test to go through all of the levels. Back to @lbergelson .

lbergelson commented:
Thanks for the changes.  Sorry to keep poking at this, but since we're mapping these enums back and fourth in several places, could we keep track of the mapping explicitly in a map?  Guava provides and [`EnumBiMap`](http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/collect/EnumBiMap.html) which seems perfect for this situation.

lbergelson commented:
@cmnbroad Or maybe it's silly to bother with since we're moving all the logging to log4j anyway.

cmnbroad commented:
Its not a bad idea; lets hold off on taking this until we're ready to take #612. At that point, we'll have to decide if we want to continue to use the samtools Log.Level enum as the type for the VERBOSITY args. That made sense in the past, but once we have no dependency on the picard logger its less compelling. However, we can't just use the log4j one since its not an enum (the values are represented by a bunch of static objects) and it doesn't have a constructor that conforms to the command line arg processors requirements. So the choices would be to retain the dependency on the picard enum, or write our own enum for HB and  then write the mapping code you suggest between that and log4j (which I'd vote for).

In the meantime I've moved and updated the tests as you suggested.

cmnbroad commented:
@lbergelson Let me know what you think of my last comment above, then reassign back to me.

lbergelson commented:
Closing this pull request since it's moved to #787
